[CI] Failure in org.elasticsearch.upgrades.BasicLicenseUpgradeIT.testOldAndMixedClusterHaveActiveBasic

```
./gradlew ':x-pack:qa:rolling-upgrade-basic:v7.9.4#oldClusterTest' --tests "org.elasticsearch.upgrades.BasicLicenseUpgradeIT.testOldAndMixedClusterHaveActiveBasic" -Dtests.seed=5B22100B1FCA9959 -Dtests.security.manager=true -Dtests.locale=hr-HR -Dtests.timezone=Africa/Algiers -Druntime.java=8
```

failed in https://gradle-enterprise.elastic.co/s/7n4cfvbfcywl6/console-log?anchor=12279 with

```

org.elasticsearch.upgrades.BasicLicenseUpgradeIT > testOldAndMixedClusterHaveActiveBasic FAILED |  
-- | --
  | org.elasticsearch.client.ResponseException: method [GET], host [http://[::1]:42229], URI [/_license], status line [HTTP/1.1 404 Not Found] |  
  | { } |  
  | at __randomizedtesting.SeedInfo.seed([5B22100B1FCA9959:78F3B9C5F9CAB7D8]:0) |  
  | at org.elasticsearch.client.RestClient.convertResponse(RestClient.java:318) |  
  | at org.elasticsearch.client.RestClient.performRequest(RestClient.java:288) |  
  | at org.elasticsearch.client.RestClient.performRequest(RestClient.java:262) |  
  | at org.elasticsearch.upgrades.BasicLicenseUpgradeIT.checkBasicLicense(BasicLicenseUpgradeIT.java:27) |  
  | at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:991) |  
  | at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:976) |  
  | at org.elasticsearch.upgrades.BasicLicenseUpgradeIT.testOldAndMixedClusterHaveActiveBasic(BasicLicenseUpgradeIT.java:17)
```


Does not reproduce locally.

I had a look in more detail at this.  The ML error message "Cannot log to named pipe /var/lib/jenkins/workspace/elastic+elasticsearch+7.10+multijob-unix-compatibility/os/fedora-32&&immutable/x-pack/qa/rolling-upgrade-basic/build/testclusters/v7.9.4-1/tmp/controller_log_556925 as it could not be opened for writing" is highly misleading.  What has actually happened is that the ES JVM has been killed, then the ML controller process has noticed and also exited.  To do this the call to connect the named pipe used for logging had to be interrupted.  So the only reason this pipe didn't open was that opening it was deliberately interrupted.  The second ML message, "Parent process died - ML controller exiting", is the one that is accurate.  I have opened elastic/ml-cpp#1591 to stop the confusing message getting logged in the future.

None of this is related to the root cause of the problem though.  Looking at that, the nodes in the cluster are logging errors like:

```
org.elasticsearch.transport.RemoteTransportException: [v7.10.1-0][127.0.0.1:45415][internal:cluster/coordination/join]
Caused by: org.elasticsearch.cluster.coordination.CoordinationStateRejectedException: incoming term 1 does not match current term 2
```

and:

```
org.elasticsearch.transport.RemoteTransportException: [v7.10.1-0][127.0.0.1:45415][internal:cluster/coordination/join]
Caused by: org.elasticsearch.cluster.coordination.FailedToCommitClusterStateException: node is no longer master for term 1 while handling publication
```

and:

```
org.elasticsearch.transport.RemoteTransportException: [v7.10.1-1][127.0.0.1:39947][internal:cluster/request_pre_vote]
Caused by: org.elasticsearch.cluster.coordination.CoordinationStateRejectedException: rejecting PreVoteRequest{sourceNode={v7.10.1-0}{RjSXzJ1CR2ubwHojpIeNUQ}{QC9rPeJHS-CW0042b0rgWg}{127.0.0.1}{127.0.0.1:45415}{cdhimrstw}{testattr=test, xpack.installed=true, transform.node=true}, currentTerm=2} as there is already a leader
```

So it seems that the cluster is failing to form, then eventually gets killed by the test harness, then after that the ML controllers (correctly) decide they should no longer be running and exit.

Does anyone from @elastic/es-distributed have an idea why the cluster doesn't form sometimes?  I am not sure why Fedora 32 would make a difference here.

Another: https://gradle-enterprise.elastic.co/s/bazoexvcxmrzc/tests/:x-pack:qa:rolling-upgrade-basic:v7.10.1%23oldClusterTest/org.elasticsearch.upgrades.BasicLicenseUpgradeIT/testOldAndMixedClusterHaveActiveBasic#1

Also Fedora-32, also cluster formation issues.
```
[WARN ][o.e.c.s.MasterService    ] [v7.10.1-0] failing [elected-as-master ([2] nodes joined)[{v7.10.1-0}{6Vcnc_NlToy_fddvOceKPw}{X46o4X6wQpSnEQ8rZUpIiQ}{127.0.0.1}{127.0.0.1:41405}{cdhimrstw}{xpack.installed=true, transform.node=true, testattr=test} elect leader, {v7.10.1-1}{OJpfmS-1Q5Ohc0efJqIRwQ}{019TQIXMSaCx84d8uL_NnQ}{127.0.0.1}{127.0.0.1:42367}{cdhimrstw}{testattr=test, xpack.installed=true, transform.node=true} elect leader, _BECOME_MASTER_TASK_, _FINISH_ELECTION_]]: failed to commit cluster state version [1]	
»  org.elasticsearch.cluster.coordination.FailedToCommitClusterStateException: publication failed	
»  	at org.elasticsearch.cluster.coordination.Coordinator$CoordinatorPublication$4.onFailure(Coordinator.java:1467) ~[elasticsearch-7.10.1-SNAPSHOT.jar:7.10.1-SNAPSHOT]
```

```
 Caused by: org.elasticsearch.ElasticsearchException: publication cancelled before committing: become candidate: joinLeaderInTerm	
»  	at org.elasticsearch.cluster.coordination.Publication.cancel(Publication.java:86) ~[elasticsearch-7.10.1-SNAPSHOT.jar:7.10.1-SNAPSHOT]
```

```
» [2020-12-08T02:25:07,862][INFO ][o.e.c.c.JoinHelper       ] [v7.10.1-0] failed to join {v7.10.1-2}{uG7wo3t5RXmOJ9kOW2R5nA}{Cp_fVfMKSPWvk5W9Odg8jw}{127.0.0.1}{127.0.0.1:38647}{cdhimrstw}{testattr=test, xpack.installed=true, transform.node=true} with JoinRequest{sourceNode={v7.10.1-0}{6Vcnc_NlToy_fddvOceKPw}{X46o4X6wQpSnEQ8rZUpIiQ}{127.0.0.1}{127.0.0.1:41405}{cdhimrstw}{xpack.installed=true, transform.node=true, testattr=test}, minimumTerm=4, optionalJoin=Optional.empty}	
»  org.elasticsearch.transport.RemoteTransportException: [v7.10.1-2][127.0.0.1:38647][internal:cluster/coordination/join]	
»  Caused by: org.elasticsearch.cluster.coordination.CoordinationStateRejectedException: join target is a follower	
»  	at org.elasticsearch.cluster.coordination.JoinHelper$FollowerJoinAccumulator.handleJoinRequest(JoinHelper.java:445) ~[elasticsearch-7.10.1-SNAPSHOT.jar:7.10.1-SNAPSHOT]
```

```
[WARN ][o.e.g.DanglingIndicesState] [v7.10.1-1] gateway.auto_import_dangling_indices is disabled, dangling indices will not be automatically detected or imported and must be managed manually	
» WARN ][o.e.d.FileBasedSeedHostsProvider] [v7.10.1-1] expected, but did not find, a dynamic hosts list at [/var/lib/jenkins/workspace/elastic+elasticsearch+7.x+multijob-unix-compatibility/os/fedora-32&&immutable/x-pack/qa/rolling-upgrade-basic/build/testclusters/v7.10.1-1/config/unicast_hosts.txt]	
»   ↑ repeated 2 times ↑	
» WARN ][o.e.c.s.MasterService    ] [v7.10.1-1] failing [elected-as-master ([2] nodes joined)[{v7.10.1-0}{6Vcnc_NlToy_fddvOceKPw}{X46o4X6wQpSnEQ8rZUpIiQ}{127.0.0.1}{127.0.0.1:41405}{cdhimrstw}{testattr=test, xpack.installed=true, transform.node=true} elect leader, {v7.10.1-1}{OJpfmS-1Q5Ohc0efJqIRwQ}{019TQIXMSaCx84d8uL_NnQ}{127.0.0.1}{127.0.0.1:42367}{cdhimrstw}{xpack.installed=true, transform.node=true, testattr=test} elect leader, _BECOME_MASTER_TASK_, _FINISH_ELECTION_]]: failed to commit cluster state version [1]	
»  org.elasticsearch.cluster.coordination.FailedToCommitClusterStateException: node is no longer master for term 1 while handling publication
```

@elastic/es-distributed What can we do to track this down?

--- a/x-pack/qa/rolling-upgrade-basic/src/test/java/org/elasticsearch/upgrades/BasicLicenseUpgradeIT.java
+++ b/x-pack/qa/rolling-upgrade-basic/src/test/java/org/elasticsearch/upgrades/BasicLicenseUpgradeIT.java
@@ -7,6 +7,7 @@ package org.elasticsearch.upgrades;
 
 import org.elasticsearch.client.Request;
 import org.elasticsearch.client.Response;
+import org.elasticsearch.client.ResponseException;
 
 import java.util.Map;
 
@@ -24,20 +25,30 @@ public class BasicLicenseUpgradeIT extends AbstractUpgradeTestCase {
 
     @SuppressWarnings("unchecked")
     private void checkBasicLicense() throws Exception {
-        Response licenseResponse = client().performRequest(new Request("GET", "/_license"));
-        Map<String, Object> licenseResponseMap = entityAsMap(licenseResponse);
-        Map<String, Object> licenseMap = (Map<String, Object>) licenseResponseMap.get("license");
-        assertEquals("basic", licenseMap.get("type"));
-        assertEquals("active", licenseMap.get("status"));
+        try {
+            Response licenseResponse = client().performRequest(new Request("GET", "/_license"));
+            Map<String, Object> licenseResponseMap = entityAsMap(licenseResponse);
+            Map<String, Object> licenseMap = (Map<String, Object>) licenseResponseMap.get("license");
+            assertEquals("basic", licenseMap.get("type"));
+            assertEquals("active", licenseMap.get("status"));
+        } catch (ResponseException e) {
+            // assertBusy only retries the code block if an AssertionError is thrown
+            throw new AssertionError("Failed to get license information", e);
+        }
     }
 
     @SuppressWarnings("unchecked")
     private void checkNonExpiringBasicLicense() throws Exception {
-        Response licenseResponse = client().performRequest(new Request("GET", "/_license"));
-        Map<String, Object> licenseResponseMap = entityAsMap(licenseResponse);
-        Map<String, Object> licenseMap = (Map<String, Object>) licenseResponseMap.get("license");
-        assertEquals("basic", licenseMap.get("type"));
-        assertEquals("active", licenseMap.get("status"));
-        assertNull(licenseMap.get("expiry_date_in_millis"));
+        try {
+            Response licenseResponse = client().performRequest(new Request("GET", "/_license"));
+            Map<String, Object> licenseResponseMap = entityAsMap(licenseResponse);
+            Map<String, Object> licenseMap = (Map<String, Object>) licenseResponseMap.get("license");
+            assertEquals("basic", licenseMap.get("type"));
+            assertEquals("active", licenseMap.get("status"));
+            assertNull(licenseMap.get("expiry_date_in_millis"));
+        } catch (ResponseException e) {
+            // assertBusy only retries the code block if an AssertionError is thrown
+            throw new AssertionError("Failed to get license information", e);
+        }
     }
 }

--- a/x-pack/qa/rolling-upgrade-basic/src/test/java/org/elasticsearch/upgrades/BasicLicenseUpgradeIT.java
+++ b/x-pack/qa/rolling-upgrade-basic/src/test/java/org/elasticsearch/upgrades/BasicLicenseUpgradeIT.java
@@ -7,7 +7,6 @@ package org.elasticsearch.upgrades;
 
 import org.elasticsearch.client.Request;
 import org.elasticsearch.client.Response;
-import org.elasticsearch.client.ResponseException;
 
 import java.util.Map;
 
@@ -25,30 +24,30 @@ public class BasicLicenseUpgradeIT extends AbstractUpgradeTestCase {
 
     @SuppressWarnings("unchecked")
     private void checkBasicLicense() throws Exception {
-        try {
-            Response licenseResponse = client().performRequest(new Request("GET", "/_license"));
-            Map<String, Object> licenseResponseMap = entityAsMap(licenseResponse);
-            Map<String, Object> licenseMap = (Map<String, Object>) licenseResponseMap.get("license");
-            assertEquals("basic", licenseMap.get("type"));
-            assertEquals("active", licenseMap.get("status"));
-        } catch (ResponseException e) {
-            // assertBusy only retries the code block if an AssertionError is thrown
-            throw new AssertionError("Failed to get license information", e);
-        }
+        final Request request = new Request("GET", "/_license");
+        // This avoids throwing a ResponseException when the license is not ready yet
+        // allowing to retry the check using assertBusy
+        request.addParameter("ignore", "404");
+        Response licenseResponse = client().performRequest(request);
+        assertOK(licenseResponse);
+        Map<String, Object> licenseResponseMap = entityAsMap(licenseResponse);
+        Map<String, Object> licenseMap = (Map<String, Object>) licenseResponseMap.get("license");
+        assertEquals("basic", licenseMap.get("type"));
+        assertEquals("active", licenseMap.get("status"));
     }
 
     @SuppressWarnings("unchecked")
     private void checkNonExpiringBasicLicense() throws Exception {
-        try {
-            Response licenseResponse = client().performRequest(new Request("GET", "/_license"));
-            Map<String, Object> licenseResponseMap = entityAsMap(licenseResponse);
-            Map<String, Object> licenseMap = (Map<String, Object>) licenseResponseMap.get("license");
-            assertEquals("basic", licenseMap.get("type"));
-            assertEquals("active", licenseMap.get("status"));
-            assertNull(licenseMap.get("expiry_date_in_millis"));
-        } catch (ResponseException e) {
-            // assertBusy only retries the code block if an AssertionError is thrown
-            throw new AssertionError("Failed to get license information", e);
-        }
+        final Request request = new Request("GET", "/_license");
+        // This avoids throwing a ResponseException when the license is not ready yet
+        // allowing to retry the check using assertBusy
+        request.addParameter("ignore", "404");
+        Response licenseResponse = client().performRequest(request);
+        assertOK(licenseResponse);
+        Map<String, Object> licenseResponseMap = entityAsMap(licenseResponse);
+        Map<String, Object> licenseMap = (Map<String, Object>) licenseResponseMap.get("license");
+        assertEquals("basic", licenseMap.get("type"));
+        assertEquals("active", licenseMap.get("status"));
+        assertNull(licenseMap.get("expiry_date_in_millis"));
     }
 }

